File: java/Day10.java
import java.io.*;
import java.util.*;

public class Day10 {
    public static void main(String[] args) throws FileNotFoundException {
        Scanner in = new Scanner(new FileReader("input/day10.txt"));
        List<String> commands = new ArrayList<>();
        while (in.hasNext()) commands.add(in.nextLine());
        
        System.out.println("part 1: " + CPU(commands)); // 16880
        System.out.println("part 2: ");
        CRT(commands);
    }
    
    static String line;
    static int drawing;
    static int X;
    
    private static void CRT(List<String> list) {
        X = 1;
        drawing = 0;
        line = "#";
        for (String s : list) {
            String[] curr = s.split(" ");
            checkPosition();
            if (curr[0].equals("addx")) {
                X += Integer.parseInt(curr[1]);
                checkPosition();
            }
        }
    }
    
    // for CRT
    private static void checkPosition() {
        drawing++;
        if (drawing == 40) {
            System.out.println(line);
            line = "";
            drawing = 0;
        }
        line += (X - 1 <= drawing && drawing <= X + 1) ? "#" : ".";
    }
    
    private static int CPU(List<String> list) {
        int cycle = 1;
        X = 1;
        int sum = 0;
        for (String s : list) {
            String[] curr = s.split(" ");
            cycle++;
            if (cycle % 40 == 20) sum += cycle * X;
            if (curr[0].equals("addx")) {
                X += Integer.parseInt(curr[1]);
                cycle++;
                if (cycle % 40 == 20) sum += cycle * X;
            }
        }
        return sum;
    }
}

File: py/Day10.py
# part 1
def cpu(lines):
    cycle, x, sum = 1, 1, 0
    for s in lines:
        curr = s.split(" ")
        cycle += 1
        sum += (cycle % 40 == 20) * cycle * x
        if curr[0] == "addx":
            x += int(curr[1])
            cycle += 1
            sum += (cycle % 40 == 20) * cycle * x
    return sum

# part 2
def crt(lines):
    x, drawing, line = 1, 0, "#"
    for s in lines:
        curr = s.split(" ")
        x, drawing, line = check_position(drawing, line, x)
        if curr[0] == "addx":
            x += int(curr[1])
            x, drawing, line = check_position(drawing, line, x)

def check_position(drawing, line, x):
    drawing += 1
    if drawing == 40:
        print(line)
        line, drawing = "", 0
    line += "#" if x - 1 <= drawing <= x + 1 else "."
    return [x, drawing, line]

def main():
    with open('../input/day10.txt') as file:
        lines = [line.strip() for line in file.readlines()]
        print(cpu(lines))
        crt(lines)

if __name__ == "__main__":
    main()
